src/oscar/apps/basket/abstract_models.py

```diff
@@ -924,35 +950,46 @@ def line_price_excl_tax(self):
 
     @property
     def line_price_excl_tax_incl_discounts(self):
-        if self._discount_excl_tax and self.line_price_excl_tax is not None:
-            return max(0, self.line_price_excl_tax - self._discount_excl_tax)
-        if self._discount_incl_tax and self.line_price_incl_tax is not None:
-            # This is a tricky situation.  We know the discount as calculated
-            # against tax inclusive prices but we need to guess how much of the
-            # discount applies to tax-exclusive prices.  We do this by
-            # assuming a linear tax and scaling down the original discount.
+        if self.line_price_excl_tax is None:
+            return None
+
+        excl_tax_discounts = self.discounts.excl_tax
+        incl_tax_discounts = self.discounts.incl_tax
+
+        if excl_tax_discounts and self.line_price_excl_tax is not None:
+            return max(0, self.line_price_excl_tax - excl_tax_discounts)
+
+        if incl_tax_discounts and self.line_price_incl_tax is not None:
             return max(
                 0,
                 self.line_price_excl_tax
-                - round_half_up(self._tax_ratio * self._discount_incl_tax),
+                - round_half_up(self._tax_ratio * incl_tax_discounts),
             )
+
         return self.line_price_excl_tax
 
     @property
     def line_price_incl_tax_incl_discounts(self):
+        excl_tax_discounts = self.discounts.excl_tax
+        incl_tax_discounts = self.discounts.incl_tax
+
         # We use whichever discount value is set.  If the discount value was
         # calculated against the tax-exclusive prices, then the line price
         # including tax
-        if self.line_price_incl_tax is not None and self._discount_incl_tax:
-            return max(0, self.line_price_incl_tax - self._discount_incl_tax)
-        elif self.line_price_excl_tax is not None and self._discount_excl_tax:
-            return max(
-                0,
-                round_half_up(
-                    (self.line_price_excl_tax - self._discount_excl_tax)
-                    / self._tax_ratio
-                ),
+        if self.line_price_incl_tax is not None and incl_tax_discounts:
+            return max(0, self.line_price_incl_tax - incl_tax_discounts)
+        elif self.line_price_excl_tax is not None and excl_tax_discounts:
```

Comment:
This is unfinshed, it must be checked if the rounding is done at the correct point.

Reply:
Checked and concluded rounding is not done at the correct point. To fix this issue, oscar should use 4 decimals for all prices internally. Added hint for developers to implement _tax_ratio to get rid of rounding errors.